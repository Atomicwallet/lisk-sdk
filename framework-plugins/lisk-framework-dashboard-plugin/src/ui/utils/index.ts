/*
 * Copyright Â© 2021 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { apiClient } from '@liskhq/lisk-client';
import { Block, Transaction } from '../types';

const MAX_BLOCKS = 103 * 3;
const MAX_TRANSACTIONS = 150;

export const updateStatesOnNewBlock = (
	client: apiClient.APIClient,
	newBlockStr: string,
	blocks: Block[],
	confirmedTransactions: Transaction[],
	unconfirmedTransactions: Transaction[],
): {
	blocks: Block[];
	confirmedTransactions: Transaction[];
	unconfirmedTransactions: Transaction[];
} => {
	const newBlock = client.block.toJSON(client.block.decode(newBlockStr));
	const newBlocks = [...blocks, newBlock].slice(-1 * MAX_BLOCKS) as Block[];

	for (const trs of ((newBlock as unknown) as Block).payload) {
		confirmedTransactions.push(trs);
	}
	const confirmedTransactionsIds = confirmedTransactions.map(t => t.id);
	const newUnconfirmedTransactions = unconfirmedTransactions
		.filter(t => !confirmedTransactionsIds.includes(t.id))
		.slice(-1 * MAX_TRANSACTIONS);

	const newConfirmedTransactions = confirmedTransactions.slice(-1 * MAX_TRANSACTIONS);

	return {
		blocks: newBlocks,
		confirmedTransactions: newConfirmedTransactions,
		unconfirmedTransactions: newUnconfirmedTransactions,
	};
};

export const updateStatesOnNewTransaction = (
	client: apiClient.APIClient,
	newTransactionStr: string,
	unconfirmedTransactions: Transaction[],
): Transaction[] => {
	const transaction = client.transaction.toJSON(client.transaction.decode(newTransactionStr));
	return [...unconfirmedTransactions, transaction].slice(-1 * MAX_TRANSACTIONS) as Transaction[];
};
