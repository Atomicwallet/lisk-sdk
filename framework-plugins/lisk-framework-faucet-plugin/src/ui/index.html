<html lang='en'>

<head>
    <meta charset='UTF-8'>
    <meta name='viewport' content='width=device-width, initial-scale=1.0'>
    <meta http-equiv='X-UA-Compatible' content='ie=edge'>
    <title>React Boilerplate</title>
    <meta name="description" content="React boilerplate">
    <link rel='icon' href='/public/favicon.ico'>

    <!-- initial state -->
    <script>
        var initial_state = null;
    </script>
</head>

<body style="background-color: #eee;">
    <div id="app"></div>
</body>

</html>
