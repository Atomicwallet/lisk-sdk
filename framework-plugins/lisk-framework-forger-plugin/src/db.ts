/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { KVStore } from '@liskhq/lisk-db';
import { codec } from '@liskhq/lisk-codec';
import * as os from 'os';
import { join } from 'path';
import { ensureDir } from 'fs-extra';
import { DB_KEY_FORGER_SYNC_INFO } from './constants';
import { forgerInfoSchema, forgerSyncSchema } from './schema';
import { ForgetSyncInfo } from './types';

export const getDBInstance = async (
	dataPath: string,
	dbName = 'lisk-framework-forger-plugin.db',
): Promise<KVStore> => {
	const resolvedPath = dataPath.replace('~', os.homedir());
	const dirPath = join(resolvedPath, dbName);
	await ensureDir(dirPath);

	return new KVStore(dirPath);
};

export const getForgerSyncInfo = async (db: KVStore): Promise<ForgetSyncInfo> => {
	const encodedSyncInfo = await db.get(DB_KEY_FORGER_SYNC_INFO);
	return codec.decode<ForgetSyncInfo>(forgerInfoSchema, encodedSyncInfo);
};

export const setForgerSyncInfo = async (db: KVStore, blockHeight: number): Promise<void> => {
	const encodedSyncInfo = codec.encode(forgerSyncSchema, { syncUptoHeight: blockHeight });
	await db.put(DB_KEY_FORGER_SYNC_INFO, encodedSyncInfo);
};
