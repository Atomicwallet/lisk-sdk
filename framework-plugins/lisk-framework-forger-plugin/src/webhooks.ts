/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import axios, { AxiosError } from 'axios';
import * as Debug from 'debug';

import { Webhook } from './types';
// eslint-disable-next-line new-cap
const debug = Debug('plugin:forger:webhooks');

interface HttpHeaders {
	[key: string]: string;
}

interface BlockCreated {
	readonly height: number;
	readonly reward: string;
	readonly forgerAddress: string;
}

interface BlockMissed {
	readonly height: number;
	readonly missedBlocksByAddress: { [key: string]: number };
}

interface NodeStatusChange {
	readonly reason: string;
}

interface WebHookPayload {
	readonly event: string;
	readonly timestamp: number;
	readonly payload: BlockCreated | BlockMissed | NodeStatusChange;
}

export class Webhooks {
	private readonly headers: HttpHeaders;
	private readonly registeredEvents: readonly Webhook[];

	public constructor(defaultHeaders: HttpHeaders, configuredEvents: readonly Webhook[]) {
		this.headers = defaultHeaders;
		this.registeredEvents = configuredEvents;
	}

	public async execute(eventData: WebHookPayload, targetURL: string): Promise<object> {
		return axios.post(targetURL, eventData, { headers: this.headers });
	}

	public async handleEvent(data: WebHookPayload): Promise<void> {
		const requiredEvents = [];
		for (const aRegisteredEvent of this.registeredEvents) {
			if (aRegisteredEvent.events.includes(data.event)) {
				requiredEvents.push({ url: aRegisteredEvent.url, data });
			}
		}

		for (const anEvent of requiredEvents) {
			try {
				await this.execute(anEvent.data, anEvent.url);
			} catch (err) {
				debug('Error during webhook processing', err, (err as AxiosError).response?.data);
			}
		}
	}
}
