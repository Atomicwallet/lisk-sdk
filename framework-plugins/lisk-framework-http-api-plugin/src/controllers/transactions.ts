/*
 * Copyright © 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { isBase64String } from '@liskhq/lisk-validator';
import { Request, Response } from 'express';
import { BaseChannel, PluginCodec } from 'lisk-framework';

export const getTransaction = (
	channel: BaseChannel,
	codec: PluginCodec,
) => async (req: Request, res: Response): Promise<void> => {
	const transactionId = req.params.id;

	// 400 - Client Side Error
	if (!transactionId || !isBase64String(transactionId)) {
		res.status(400).send({
			errors: [
				{ message: 'The transaction id parameter should be a base64 string.' },
			],
		});
		return;
	}

	let transaction: string;

	try {
		transaction = await channel.invoke<string>('app:getTransactionByID', {
			id: transactionId,
		});
	} catch (error) {
		if (
			(error as Error).message ===
			'Specified key transactions:id:×m does not exist'
		) {
			// 404 - Not Found Error
			res.status(404).send({
				errors: [
					{ message: `The transaction with id ${transactionId} not found.` },
				],
			});
			return;
		}

		// 500 - Server Error
		throw error;
	}

	// 200 - Response
	res.status(200).json(codec.decodeTransaction(transaction));
};
