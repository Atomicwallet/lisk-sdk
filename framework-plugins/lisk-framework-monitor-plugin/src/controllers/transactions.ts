/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { Request, Response } from 'express';
import { BaseChannel } from 'lisk-framework';
import { PeerInfo, SharedState } from '../types';

const getAverage = (
	transactionsStats: Pick<SharedState['transactions'], 'transactions'>,
): number => {
	let transactionCount = 0;
	let total = 0;

	for (const transactionStats of Object.values(transactionsStats.transactions)) {
		transactionCount += 1;
		total += transactionStats.count;
	}

	return total / transactionCount;
};

export const getTransactionStats = (channel: BaseChannel, state: SharedState) => async (
	_req: Request,
	res: Response,
): Promise<void> => {
	const { transactions } = state;
	res.json({
		data: {
			transactions: transactions.transactions,
			connectedPeers: (await channel.invoke<ReadonlyArray<PeerInfo>>('app:getConnectedPeers'))
				.length,
			averageReceivedTransactions: getAverage(transactions),
		},
		meta: {},
	});
};
