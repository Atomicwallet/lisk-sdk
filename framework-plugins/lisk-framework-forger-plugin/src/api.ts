/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import * as express from 'express';
import { Express } from 'express';
import { BaseChannel, PluginCodec } from 'lisk-framework';
import { KVStore } from '@liskhq/lisk-db';
import * as cors from 'cors';
import * as rateLimit from 'express-rate-limit';
import * as middlewares from './middlewares';
import * as controllers from './controllers';
import { Options } from './types';

export const initApi = (
	options: Options,
	channel: BaseChannel,
	codec: PluginCodec,
	db: KVStore,
): Express => {
	const app: Express = express();

	// Register before middleware
	app.use(cors(options.cors));
	app.use(express.json());
	app.use(rateLimit(options.limits));
	app.use(middlewares.whiteListMiddleware(options));

	// Register controllers
	app.get('/api/voters', controllers.voters.getVoters(channel, codec, db));
	app.patch('/api/forging', controllers.forging.updateForging(channel, db));
	app.get('/api/forging/info', controllers.forgingInfo.getForgingInfo(channel, codec, db));

	// Register after middleware
	app.use(middlewares.errorMiddleware());

	return app;
};
