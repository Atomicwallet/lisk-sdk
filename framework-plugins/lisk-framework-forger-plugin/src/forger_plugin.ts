/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { Server } from 'http';
import * as Debug from 'debug';
import { getAddressFromPublicKey } from '@liskhq/lisk-cryptography';
import { codec } from '@liskhq/lisk-codec';
import { KVStore } from '@liskhq/lisk-db';
import {
	ActionsDefinition,
	BasePlugin,
	BaseChannel,
	EventsArray,
	EventInfoObject,
	PluginInfo,
	TransactionJSON,
	BlockHeaderJSON,
} from 'lisk-framework';
import { VoteTransaction } from '@liskhq/lisk-transactions';
import { objects } from '@liskhq/lisk-utils';
import type { Express } from 'express';
import { initApi } from './api';
import { getDBInstance, getForgerSyncInfo } from './db';
import * as config from './defaults';
import { Forger, ForgerInfo, Options, TransactionFees } from './types';
import { DB_KEY_FORGER_INFO } from './constants';
import { forgerInfoSchema } from './schema';

interface Data {
	readonly block: string;
}

interface Asset {
	readonly votes: Array<Readonly<Vote>>;
}
interface Vote {
	delegateAddress: string;
	amount: string;
}

interface ForgerPayloadInfo {
	forgerAddress: string;
	forgerAddressBinary: string;
	header: BlockHeaderJSON;
	payload: readonly TransactionJSON[];
}

interface NodeInfo {
	genesisConfig: {
		readonly blockTime: number;
	};
}

// eslint-disable-next-line
const packageJSON = require('../package.json');
// eslint-disable-next-line new-cap
const debug = Debug('plugin:forger');

export class ForgerPlugin extends BasePlugin {
	private _forgerPluginDB!: KVStore;
	private _server!: Server;
	private _app!: Express;
	private _channel!: BaseChannel;
	private _forgersList!: ReadonlyArray<Forger>;
	private _transactionFees!: TransactionFees;

	// eslint-disable-next-line @typescript-eslint/class-literal-property-style
	public static get alias(): string {
		return 'forger';
	}

	// eslint-disable-next-line @typescript-eslint/class-literal-property-style
	public static get info(): PluginInfo {
		return {
			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
			author: packageJSON.author,
			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
			version: packageJSON.version,
			// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
			name: packageJSON.name,
		};
	}

	// eslint-disable-next-line class-methods-use-this
	public get defaults(): object {
		return config.defaultConfig;
	}

	// eslint-disable-next-line class-methods-use-this
	public get events(): EventsArray {
		return [];
	}

	// eslint-disable-next-line class-methods-use-this
	public get actions(): ActionsDefinition {
		return {};
	}

	// eslint-disable-next-line @typescript-eslint/require-await
	public async load(channel: BaseChannel): Promise<void> {
		const options = objects.mergeDeep({}, config.defaultConfig.default, this.options) as Options;
		this._channel = channel;

		this._forgerPluginDB = await getDBInstance(options.dataPath);

		// eslint-disable-next-line @typescript-eslint/no-misused-promises
		this._channel.once('app:ready', async () => {
			this._app = initApi(options, this._channel, this.codec);

			// Fetch and set forger list from the app
			await this._setForgersList();

			// Fetch and set transactions fees
			await this._setTransactionFees();

			// Sync the information
			await this._syncForgerInfo();

			// Listen to new block and delete block events
			this._subscribeToChannel();

			// Start http server
			this._server = this._app.listen(options.port, '0.0.0.0');
		});
	}

	public async unload(): Promise<void> {
		await new Promise((resolve, reject) => {
			this._server.close(err => {
				if (err) {
					reject(err);
					return;
				}
				resolve();
			});
		});

		await this._forgerPluginDB.close();
	}

	private async _syncForgerInfo(): Promise<void> {
		const lastBlock = this.codec.decodeBlock(
			await this._channel.invoke<Buffer>('app:getLastBlock'),
		);
		const forgerSyncInfo = await getForgerSyncInfo(this._forgerPluginDB);

		if (forgerSyncInfo.syncUptoHeight === lastBlock.header.height) {
			// No need to sync
			return;
		}

		let syncFromHeight: number;

		if (forgerSyncInfo.syncUptoHeight > lastBlock.header.height) {
			await this._forgerPluginDB.clear();
			syncFromHeight = 1;
		} else {
			syncFromHeight = lastBlock.header.height + 1;
		}

		const blocks = await this._channel.invoke<string[]>('app: getBlocksByHeightBetween', {
			from: syncFromHeight,
			to: lastBlock.header.height,
		});

		for (const block of blocks) {
			await this._incrementForgerInfo(block);
		}

		// Try to sync again if more blocks forged meanwhile
		await this._syncForgerInfo();
	}

	private _subscribeToChannel(): void {
		// eslint-disable-next-line @typescript-eslint/no-misused-promises
		this._channel.subscribe('app:block:new', async (eventInfo: EventInfoObject) => {
			const { block } = eventInfo.data as Data;
			await this._incrementForgerInfo(block);
		});

		// eslint-disable-next-line @typescript-eslint/no-misused-promises
		this._channel.subscribe('app:block:delete', async (eventInfo: EventInfoObject) => {
			const { block } = eventInfo.data as Data;
			await this._decrementForgerInfo(block);
		});
	}

	private async _setForgersList(): Promise<void> {
		this._forgersList = await this._channel.invoke<Forger[]>('app:getForgingStatusOfAllDelegates');
	}

	private async _getForgerInfo(forgerAddress: string): Promise<ForgerInfo> {
		let forgerInfo;
		try {
			forgerInfo = await this._forgerPluginDB.get(`${DB_KEY_FORGER_INFO}:${forgerAddress}`);
		} catch (error) {
			debug(`Forger info does not exists for delegate: ${forgerAddress}`);
			return {
				totalProducedBlocks: 0,
				totalMissedBlocks: 0,
				totalReceivedFees: BigInt(0),
				totalReceivedRewards: BigInt(0),
				votesReceived: [],
			};
		}

		return codec.decode<ForgerInfo>(forgerInfoSchema, forgerInfo);
	}

	private async _setForgerInfo(forgerAddress: string, forgerInfo: ForgerInfo): Promise<void> {
		const encodedForgerInfo = codec.encode(forgerInfoSchema, forgerInfo);
		await this._forgerPluginDB.put(`${DB_KEY_FORGER_INFO}:${forgerAddress}`, encodedForgerInfo);
	}

	private async _setTransactionFees(): Promise<void> {
		this._transactionFees = await this._channel.invoke<TransactionFees>('app:getTransactionsFees');
	}

	private _getForgerHeaderAndPayloadInfo(block: string): ForgerPayloadInfo {
		const { header, payload } = this.codec.decodeBlock(block);
		const forgerAddress = getAddressFromPublicKey(
			Buffer.from(header.generatorPublicKey, 'base64'),
		).toString('base64');
		const forgerAddressBuffer = Buffer.from(forgerAddress, 'base64');
		const forgerAddressBinary = forgerAddressBuffer.toString('binary');

		return {
			forgerAddress,
			forgerAddressBinary,
			header,
			payload,
		};
	}

	private async _incrementForgerInfo(encodedBlock: string): Promise<void> {
		const {
			forgerAddress,
			forgerAddressBinary,
			header: { reward },
			payload,
		} = this._getForgerHeaderAndPayloadInfo(encodedBlock);
		const forgerInfo = await this._getForgerInfo(forgerAddressBinary);
		let isUpdated = false;

		if (this._forgersList.find(forger => forger.address === forgerAddress)) {
			forgerInfo.totalProducedBlocks += 1;
			forgerInfo.totalReceivedRewards += BigInt(reward);
			forgerInfo.totalReceivedFees += this._getFee(payload, encodedBlock);
			isUpdated = true;
		}

		for (const trx of payload) {
			if (trx.type === VoteTransaction.TYPE) {
				for (const vote of (trx.asset as Asset).votes) {
					const registeredDelegateIndex = this._forgersList.findIndex(
						forger => forger.address === vote.delegateAddress,
					);
					if (registeredDelegateIndex !== -1) {
						const delegateVoteIndex = forgerInfo.votesReceived.findIndex(aVote =>
							aVote.address.equals(Buffer.from(vote.delegateAddress, 'base64')),
						);
						if (delegateVoteIndex === -1) {
							forgerInfo.votesReceived.push({
								address: Buffer.from(vote.delegateAddress, 'base64'),
								amount: BigInt(vote.amount),
							});
						} else {
							forgerInfo.votesReceived[delegateVoteIndex].amount += BigInt(vote.amount);
						}
						isUpdated = true;
					}
				}
			}
		}

		await this._updateMissedBlock(encodedBlock);

		if (isUpdated) {
			await this._setForgerInfo(forgerAddressBinary, { ...forgerInfo });
		}
	}

	private async _decrementForgerInfo(encodedBlock: string): Promise<void> {
		const {
			forgerAddress,
			forgerAddressBinary,
			header: { reward },
			payload,
		} = this._getForgerHeaderAndPayloadInfo(encodedBlock);
		const forgerInfo = await this._getForgerInfo(forgerAddressBinary);
		let isUpdated = false;

		if (this._forgersList.find(forger => forger.address === forgerAddress)) {
			forgerInfo.totalProducedBlocks -= 1;
			forgerInfo.totalReceivedRewards -= BigInt(reward);
			forgerInfo.totalReceivedFees -= this._getFee(payload, encodedBlock);
			isUpdated = true;
		}

		for (const trx of payload) {
			if (trx.type === VoteTransaction.TYPE) {
				for (const vote of (trx.asset as Asset).votes) {
					const delegateVoteIndex = forgerInfo.votesReceived.findIndex(aVote =>
						aVote.address.equals(Buffer.from(vote.delegateAddress, 'base64')),
					);
					if (delegateVoteIndex !== -1) {
						forgerInfo.votesReceived[delegateVoteIndex].amount -= BigInt(vote.amount);
						if (forgerInfo.votesReceived[delegateVoteIndex].amount === BigInt(0)) {
							forgerInfo.votesReceived.splice(delegateVoteIndex, 1);
						}
						isUpdated = true;
					}
				}
			}
		}

		if (isUpdated) {
			await this._setForgerInfo(forgerAddressBinary, { ...forgerInfo });
		}
	}

	private _getFee(payload: ReadonlyArray<TransactionJSON>, block: string): bigint {
		const { payload: payloadBuffer } = this.codec.decodeRawBlock(block);
		let fee = BigInt(0);

		for (let index = 0; index < payload.length; index += 1) {
			const trx = payload[index];
			fee +=
				BigInt(this._transactionFees[trx.type].baseFee) +
				BigInt(this._transactionFees[trx.type].minFeePerByte) * BigInt(payloadBuffer[index].length);
		}

		return fee;
	}

	private async _updateMissedBlock(block: string): Promise<void> {
		const {
			header: { height, timestamp },
			forgerAddress,
		} = this._getForgerHeaderAndPayloadInfo(block);
		const previousBlockStr = await this._channel.invoke<string>('app:getBlockByHeight', {
			height: height - 1,
		});
		const {
			genesisConfig: { blockTime },
		} = await this._channel.invoke<NodeInfo>('app:getNodeInfo');
		const { header: previousBlock } = this.codec.decodeBlock(previousBlockStr);
		const missedBlocks = Math.ceil((timestamp - previousBlock.timestamp) / blockTime) - 1;

		if (missedBlocks > 0) {
			const round = await this._channel.invoke('app:getSlotRound', { height });
			const forgerAddressForRound = await this._channel.invoke<readonly string[]>(
				'app:getForgerAddressesForRound',
				{ round },
			);
			const forgersRoundLength = forgerAddressForRound.length;
			const forgerIndex = forgerAddressForRound.findIndex(address => address === forgerAddress);

			for (let index = 0; index < missedBlocks; index += 1) {
				const rawIndex = (forgerIndex - 1 - index) % forgersRoundLength;
				const forgerRoundIndex = rawIndex >= 0 ? rawIndex : rawIndex + forgersRoundLength;
				const missedForgerAddress = forgerAddressForRound[forgerRoundIndex];
				const missedForger = await this._getForgerInfo(missedForgerAddress);
				missedForger.totalMissedBlocks += 1;
				await this._setForgerInfo(missedForgerAddress, missedForger);
			}
		}
	}
}
