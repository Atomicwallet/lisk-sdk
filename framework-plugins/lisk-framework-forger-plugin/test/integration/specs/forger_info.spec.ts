/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { Application } from 'lisk-framework';
import { getAddressFromPublicKey } from '@liskhq/lisk-cryptography';

import axios from 'axios';
import {
	closeApplication,
	createApplication,
	getURL,
	callNetwork,
	waitNBlocks,
	waitTill,
} from '../../utils/application';
import { ForgerPlugin } from '../../../src';
import { getRandomAccount } from '../../utils/accounts';
import { createTransferTransaction, createVoteTransaction } from '../../utils/transactions';

const getForgerInfo = async (forgerPluginInstance: ForgerPlugin, generatorPublicKey: string) => {
	const forgerAddress = getAddressFromPublicKey(Buffer.from(generatorPublicKey, 'base64')).toString(
		'base64',
	);
	const forgerInfo = await forgerPluginInstance['_getForgerInfo'](forgerAddress);

	return forgerInfo;
};

describe('Forger Info', () => {
	let app: Application;
	let accountNonce = 0;

	beforeAll(async () => {
		app = await createApplication('transactions');
	});

	afterAll(async () => {
		await closeApplication(app);
	});

	describe('New Block', () => {
		it('should save forger info after new block', async () => {
			// Arrange
			const forgerPluginInstance = app['_controller'].plugins[ForgerPlugin.alias];

			// Act
			const result = await axios.get(getURL('/api/blocks/?height=1'));
			const {
				data: {
					data: {
						header: { generatorPublicKey },
					},
				},
			} = result;
			const forgerInfo = await getForgerInfo(forgerPluginInstance, generatorPublicKey);

			// Assert
			expect(forgerInfo).toMatchSnapshot();
		});

		it('should save forger info with received fees if payload included in new block', async () => {
			// Arrange
			const forgerPluginInstance = app['_controller'].plugins[ForgerPlugin.alias];
			const account = getRandomAccount();
			const transaction = createTransferTransaction({
				amount: '2',
				recipientAddress: account.address,
				fee: '0.3',
				nonce: accountNonce,
			});
			accountNonce += 1;

			const { id, ...input } = transaction;
			const { response, status } = await callNetwork(
				axios.post(getURL('/api/transactions'), input),
			);
			expect(status).toEqual(200);
			expect(response).toEqual({ data: { transactionId: id }, meta: {} });
			await waitNBlocks(app, 1);

			const { response: getResponse, status: getStatus } = await callNetwork(
				// eslint-disable-next-line @typescript-eslint/restrict-template-expressions
				axios.get(getURL(`/api/transactions/${encodeURIComponent(id)}`)),
			);

			expect(getStatus).toEqual(200);
			expect(getResponse).toEqual({ data: transaction, meta: {} });

			const {
				header: { generatorPublicKey },
			} = app['_node']['_chain'].lastBlock;
			const forgerInfo = await getForgerInfo(forgerPluginInstance, generatorPublicKey);

			// Assert
			expect(forgerInfo).toMatchSnapshot();
		});

		it('should save forger info with votes received in new block', async () => {
			// Arrange
			const forgerPluginInstance = app['_controller'].plugins[ForgerPlugin.alias];
			const forgingDelegateAddress = forgerPluginInstance['_forgersList'][0].address;
			const transaction = createVoteTransaction({
				amount: '10',
				recipientAddress: forgingDelegateAddress,
				fee: '0.3',
				nonce: accountNonce,
			});
			accountNonce += 1;

			const { id, ...input } = transaction;
			const { response, status } = await callNetwork(
				axios.post(getURL('/api/transactions'), input),
			);
			expect(status).toEqual(200);
			expect(response).toEqual({ data: { transactionId: id }, meta: {} });
			await waitNBlocks(app, 1);

			const { response: getResponse, status: getStatus } = await callNetwork(
				// eslint-disable-next-line @typescript-eslint/restrict-template-expressions
				axios.get(getURL(`/api/transactions/${encodeURIComponent(id)}`)),
			);

			expect(getStatus).toEqual(200);
			expect(getResponse).toEqual({ data: transaction, meta: {} });

			const {
				header: { generatorPublicKey },
			} = app['_node']['_chain'].lastBlock;
			const forgerInfo = await getForgerInfo(forgerPluginInstance, generatorPublicKey);

			// Assert
			expect(forgerInfo).toMatchSnapshot();
		});
	});

	describe('Delete Block', () => {
		it('should update forger info after delete block', async () => {
			// Arrange
			const { generatorPublicKey } = app['_node']['_chain'].lastBlock.header;
			const forgerPluginInstance = app['_controller'].plugins[ForgerPlugin.alias];
			// const forgingDelegateAddress = forgerPluginInstance['_forgersList'][0].address;
			await app['_node']['_processor'].deleteLastBlock();

			// Act
			await waitTill(50);
			const forgerInfo = await getForgerInfo(forgerPluginInstance, generatorPublicKey);

			// Asserts
			expect(forgerInfo).toMatchSnapshot();
		});
	});
});
