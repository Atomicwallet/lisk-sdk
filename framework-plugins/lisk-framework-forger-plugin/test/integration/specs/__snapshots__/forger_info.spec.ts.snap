// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Forger Info Delete Block should update forger info after delete block 1`] = `
Object {
  "totalProducedBlocks": 0,
  "totalReceivedFees": 0n,
  "totalReceivedRewards": 0n,
  "votesReceived": Array [
    Object {
      "address": Object {
        "data": Array [
          156,
          171,
          238,
          61,
          39,
          66,
          102,
          118,
          184,
          82,
          206,
          107,
          128,
          76,
          178,
          253,
          255,
          124,
          208,
          181,
        ],
        "type": "Buffer",
      },
      "amount": 0n,
    },
  ],
}
`;

exports[`Forger Info New Block should save forger info after new block 1`] = `
Object {
  "totalProducedBlocks": 1,
  "totalReceivedFees": 0n,
  "totalReceivedRewards": 0n,
  "votesReceived": Array [],
}
`;

exports[`Forger Info New Block should save forger info with received fees if payload included in new block 1`] = `
Object {
  "totalProducedBlocks": 1,
  "totalReceivedFees": 140000n,
  "totalReceivedRewards": 0n,
  "votesReceived": Array [],
}
`;

exports[`Forger Info New Block should save forger info with votes received in new block 1`] = `
Object {
  "totalProducedBlocks": 1,
  "totalReceivedFees": 141000n,
  "totalReceivedRewards": 0n,
  "votesReceived": Array [
    Object {
      "address": Object {
        "data": Array [
          156,
          171,
          238,
          61,
          39,
          66,
          102,
          118,
          184,
          82,
          206,
          107,
          128,
          76,
          178,
          253,
          255,
          124,
          208,
          181,
        ],
        "type": "Buffer",
      },
      "amount": 1000000000n,
    },
  ],
}
`;
